# -*- coding: utf-8 -*-
#
#  This file is part of Sequana software
#
#  Copyright (c) 2016-2021 - Sequana Dev Team (https://sequana.readthedocs.io)
#
#  Distributed under the terms of the 3-clause BSD license.
#  The full license is in the LICENSE file, distributed with this software.
#
#  Website:       https://github.com/sequana/sequana
#  Documentation: http://sequana.readthedocs.io
#  Contributors:  https://github.com/sequana/sequana/graphs/contributors
##############################################################################
import sequana
from sequana import snaketools as sm
from sequana import sequana_data

# This must be defined before the include
configfile: "config.yaml"

# A convenient manager
manager = sm.PipelineManager("multitax", config)
manager.setup(globals(), mode="warning")


__taxonomy__input_fastq = manager.getrawdata()
__taxonomy__output_summary = '{sample}/kraken/kraken.out.summary'
__taxonomy__output_summary_json = '{sample}/kraken/summary.json'
__taxonomy__output_html = '{sample}/summary.html'
__taxonomy__output_csv = '{sample}/kraken/kraken.csv'


expected_output = ["outputs/dendogram.png", "multiqc/multiqc_report.html",
                    "outputs/proportion_kraken.png"]

if len(config['sequana_taxonomy']["databases"]) > 1:
    expected_output += ["outputs/proportion_dbs.png"]



rule all:
    input: 
        expected_output,
#        expand("{sample}/kraken/{sample}_unclassified_1.fastq", sample=sorted(manager.samples)),
#        expand("{sample}/kraken/{sample}_unclassified_1.fasta", sample=sorted(manager.samples)),
#        expand("{sample}/kraken/{sample}_1000_unclassified_1.fasta", sample=sorted(manager.samples)),
#        expand("{sample}/blast/{sample}_1000_unclassified_1.tsv", sample=sorted(manager.samples)),
#        expand("{sample}/blast/{sample}_1000_unclassified_1.tsv_Results.csv", sample=sorted(manager.samples)),
        expand("{sample}/blast/{sample}_1000_unclassified_1.tsv_Results.csv_Results.txt", sample=sorted(manager.samples)),
        expand("{sample}/blast/text.krona.html", sample=sorted(manager.samples))

rule sequana_taxonomy:
    input: __taxonomy__input_fastq
    output:
        html         = __taxonomy__output_html,
        csv          = __taxonomy__output_csv,
        summary      = __taxonomy__output_summary,
        summary_json = __taxonomy__output_summary_json,
        unclassified = "{sample}/kraken/{sample}_unclassified_1.fastq"
    threads: 4
    run:
        outdir = output.html.split("/",1)[0]
        if len(input) == 2:
            cmd = "sequana_taxonomy --file1 {} --file2 {} ".format(input[0], input[1])
        elif len(input) == 1:
            cmd = "sequana_taxonomy --file1 {} ".format(input[0])

        if config["sequana_taxonomy"]["confidence"] != 0:
            cmd += " --confidence {} ".format(config["sequana_taxonomy"]["confidence"])

        cmd += " --thread {} --databases".format(threads)
        for this in config["sequana_taxonomy"]['databases']:
            if this != "toydb":
                assert os.path.exists(this), "databases {} does not exits".format(this)
            cmd += " {} ".format(this)
        cmd += " --output-directory {} --level {}".format(wildcards.sample, config['sequana_taxonomy']['level'])
        if config['sequana_taxonomy']['store_unclassified']:
            if len(input) == 2:
                cmd += " --unclassified-out {}_unclassified#.fastq".format(wildcards.sample)
            else:
                cmd += " --unclassified-out {}_unclassified_1.fastq".format(wildcards.sample)
        cmd += config['sequana_taxonomy']['options']

        shell(cmd)


__summary_plot__output_summary = "outputs/sequana_kraken_summary.json"
__summary_plot__output_image = "outputs/proportion_kraken.png"


#rule kraken_2_blast:
#    input: "{sample}/kraken/{sample}_unclassified_1.fastq"
#    output: 
#        "{sample}/kraken/{sample}_unclassified_1.fasta",
#        "{sample}/kraken/{sample}_1000_unclassified_1.fasta",
#        "{sample}/blast/{sample}_1000_unclassified_1.tsv"
#    threads: 8
#    shell:
#        """
#        sed '/^@/!d;s//>/;N' {input} > {output[0]}
#        head -n2000 {output[0]} > {output[1]}
#        module load blast+/2.10.0
#        blastn -query {output[1]} -db /pasteur/zeus/projets/p02/Biomics/resources/blast/nt -num_threads {threads} -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore staxids" > {output[2]}
#        """

rule blast_analysis:
    input: "{sample}/blast/{sample}_1000_unclassified_1.tsv"
    output:
        "{sample}/blast/{sample}_analysis.csv",
        "{sample}/blast/{sample}_analysis.txt"
    run:
        from blast import remove_duplicates
        remove_duplicates(input[0]) > output[0]
        from blast import krona
        krona(output[0]) > output[1]

rule krona_visu:
    input: 
        "{sample}/blast/{sample}_analysis.txt"
    output:
        "{sample}/blast/text.krona.html"
    shell:
        """
        ktImportText {input} -o {output}
        """

rule summary_plot:
    input: expand("{sample}/kraken/kraken.csv", sample=sorted(manager.samples))
    output:
        image=__summary_plot__output_image,
        data = __summary_plot__output_summary
    run:
        from sequana.kraken import MultiKrakenResults
        k = MultiKrakenResults(input, sample_names=sorted(manager.samples))
        k.plot_stacked_hist(output.image, dpi=200,
            kind=config["summary_plot"]["kind"],
            width=config["summary_plot"]["width"],
            ytick_fontsize=config['summary_plot']["yticks_fontsize"],
            lw=config['summary_plot']["linewidth"],
            max_labels=config['summary_plot']["max_labels"],
            edgecolor=config['summary_plot']["edgecolor"],
            )
        import json
        with open(output.data, "w") as fout:
             json.dump(k.get_df().to_dict(), fout, indent=True, sort_keys=True)

if len(config["sequana_taxonomy"]["databases"]) > 1:
    __summary_plot_db__output_image = "outputs/proportion_dbs.png"
    __summary_plot_db__output_summary = "outputs/sequana_kraken_dbs_summary.json"
    rule summary_plot_db:
        input: expand(__taxonomy__output_summary_json, sample=manager.samples)
        output:
            image=__summary_plot_db__output_image,
            data = __summary_plot_db__output_summary
        run:
            from sequana.kraken import MultiKrakenResults2
            k = MultiKrakenResults2(input, manager.samples)
            if len(config['summary_plot_db']["colors"]) == 0:
                colors = None
            else:
                colors = config['summary_plot_db']["colors"]
            k.plot_stacked_hist(output.image, dpi=200,
                ytick_fontsize=config['summary_plot_db']["yticks_fontsize"],
                lw=config['summary_plot_db']["linewidth"],
                max_labels=config['summary_plot_db']["max_labels"],
                edgecolor=config['summary_plot_db']["edgecolor"],
                alpha=config['summary_plot_db']["alpha"],
                colors=colors,
                cmap=config['summary_plot_db']["colormap"])
            import json
            with open(output.data, "w") as fout:
                 json.dump(k.get_df().to_dict(), fout, indent=True, sort_keys=True)



__multiqc__input = expand("{sample}/kraken/kraken.csv", sample=sorted(manager.samples))
__multiqc__output = "multiqc/multiqc_report.html"
include: sm.modules["multiqc/2.0"]


rule dendogram:
    input: __summary_plot__output_summary
    output: "outputs/dendogram.png"
    params:
        sample_file=config["dendogram"]["sample_file"]
    run:
        data = json.loads(open(input[0], "r").read())
        import pandas as pd
        import pylab
        df = pd.DataFrame(data)
        if len(manager.samples.keys()) == 1:
            from sequana import sequana_data
            data = sequana_data("no_data.png")
            import shutil
            shutil.copy(data, output[0])
            return

        from sequana.viz.heatmap import Heatmap
        h = Heatmap(df.fillna(0))

        if params.sample_file:
            df = pd.read_csv(params.sample_file)
            codes =  pd.Categorical(df.category).codes
            df['codes'] = codes
            #for sample, category in zip(df['sample'], df['category']):
            for sample, category in zip(df['sample'], df['codes']):
                h.category_column[sample] = category
        try:
            h.plot(colorbar_position="top left")
            pylab.savefig(output[0], dpi=200)
            with open("outputs/dendogram.info", "w") as fout:
                fout.write("ok")
        except:
            from sequana import sequana_data
            data = sequana_data("no_data.png")
            import shutil
            shutil.copy(data, output[0])
            with open("outputs/dendogram.info", "w") as fout:
                fout.write("failed")





# Those rules takes a couple of seconds so no need for a cluster
localrules: multiqc,  summary_plot, dendogram 


onsuccess:

    from sequana import logger
    logger.setLevel("INFO")

    # This should create the stats plot and the Makefile
    manager.teardown()
    manager.clean_multiqc(__multiqc__output)


    # HTML report:
    from sequana.modules_report.summary import SummaryModule2
    from sequana.modules_report.kraken import KrakenModule
    intro = """<h2>General Information</h2>{}""".format(manager.get_html_summary())
    intro += """<div style="width:80%">This pipeline summarizes the taxonomic analysis
                made on {} samples. Here below you can find a summary
                of the analysis. The first plot (left) shows the proportion of
                reads classified by the databases in each found kingdom (Eukaryota,
                Bacteria, Viruses, Archea). If several databases were used,
                you should also see a plot (right panel) with proportion of reads
                classified in each database (rather than by kingdom).
                The per-sample analysis are made with kraken. A multi-sample
                report is also available as
                <a href="multiqc/multiqc_report.html">sequana multiqc</a> report.
                Individual report are also browsable (Go to the <a href="#kraken">
                Individual taxonomic
                reports</a> section here below).</div>""".format(len(manager.samples))


    intro += """<div style="clear:both"></div>"""
    intro += """<h2>Analysis overview</h2>
        <div style="width:48%; float:left"><p>Proportion of reads in virus, bacteria, human and unclassified categories is shown here below. The underlying data can be downloaded from <a href="outputs/sequana_kraken_summary.json">sequana_kraken_summary.json</a> file or here below as a CSV file.<br></p>
    """
    img = SummaryModule2.png_to_embedded_png("dummy_self", "outputs/proportion_kraken.png",
        style="max-height:60%; max-width:80%", title="Kingdom proportion (and unclassified) for each sample")
    intro += img + "</div>"
    if len(config['sequana_taxonomy']['databases'])>1:
        intro += """
            <div style="width:48%; float:right"><p>Proportion of reads found in each databases is shown here below. The underlying data can be found
in <a href="outputs/sequana_kraken_dbs_summary.json">sequana_kraken_dbs_summary.json</a> file or as a CSV file herebelow.<br></p>
        """
        img = SummaryModule2.png_to_embedded_png("dummy_self", "outputs/proportion_dbs.png",
                    style="max-height:60%; max-width:80%", title="DB proportion (and unclassified) per sample")
        intro += img + "</div>"

        from sequana.kraken import MultiKrakenResults2
        k = MultiKrakenResults2(["{}/kraken/summary.json".format(x) for x in manager.samples],
            manager.samples)
        from sequana.utils.datatables_js import DataTable
        df = k.get_df()
        links = ["""{}/summary.html""".format(sample) for sample in df.columns]
        df = df.T
        df.reset_index(inplace=True)
        df['link'] = links
        df = df.rename({"index": "sample"}, axis=1)
        datatable = DataTable(df, 'kraken', index=False)
        datatable.datatable.datatable_options = {'paging': 'false',
                                               'buttons': ['copy', 'csv'],
                                              'bSort': 'true',
                                             'dom':"BRSPfrti"
                                             }
        datatable.datatable.set_links_to_column('link', 'sample')
        js = datatable.create_javascript_function()
        htmltable = datatable.create_datatable()
        intro += """<div style="clear:both"><hr>"""+js + htmltable + "</div>"

    from sequana.kraken import MultiKrakenResults
    k = MultiKrakenResults(["{}/kraken/kraken.csv".format(x) for x in manager.samples],
        manager.samples)
    from sequana.utils.datatables_js import DataTable
    df = k.get_df()
    links = ["""{}/summary.html""".format(sample) for sample in df.columns]
    df = df.T
    df.reset_index(inplace=True)
    df['link'] = links
    df = df.rename({"index": "sample"}, axis=1)
    datatable = DataTable(df, 'kraken_kingdom', index=False)
    datatable.datatable.datatable_options = {'paging': 'false',
                                               'buttons': ['copy', 'csv'],
                                              'bSort': 'true',
                                             'dom':"BRSPfrti"
                                             }
    datatable.datatable.set_links_to_column('link', 'sample')
    js = datatable.create_javascript_function()
    htmltable = datatable.create_datatable(float_format="%0.3f")
    intro += """<div style="clear:both"><hr>"""+js + htmltable + "</div>"

    if os.path.exists("outputs/dendogram.info"):
        data = open("outputs/dendogram.info").readline()
        if data == "ok":
            image = SummaryModule2.png_to_embedded_png("self_dummy", "outputs/dendogram.png",
                style="align:center; width:45%; height:45%", title="Clustering of kingdom proportion (and unclassified")
            intro += "<h2>Results clustering (kingdom-based)</h2>"
            intro += """<div style="width:80%">From the results of the classification of reads, we extract the proportion of reads in each kingdom and clusterise the data. Here below is the results of the clusterisation. The x-axis shows the sample names, the y-axis includes the kingdom. Unclassified reads are also taken into account as part of the kingdoms. </div>"""
            intro += image

    intro +="""
    <div style="width:80%">Each of the sample ({} in total) was analysed independently using {} databases. Here below you can find the links towards each analysis. You can either click on the link or the image to open a new report. The report includes a table with all taxon found as well as their full lineage as provided by Kraken analysis tool. A dynamic plot based on Kroan allows you to explore the results interactively. </div>
    """.format(len(manager.samples), len(config["sequana_taxonomy"]["databases"]))

    intro += "<ul>"
    for sample in sorted(manager.samples):
        intro += """<li><a href="{}/summary.html">{} individual report</a></li>""".format(sample, sample)
    intro += "</ul>"

    # now include the pie images
    for sample in sorted(manager.samples):
        image = SummaryModule2.png_to_embedded_png("self_dummy",
                    "{}/kraken/kraken.png".format(sample),
                    style="align:center; width:30%; height:30%", alt=sample, title=sample)
        intro += '<a href="{}/summary.html">'.format(sample) + image + '</a>'

#s = SummaryModule2(data, intro=intro)


    shell("chmod -R g+w .")

onerror:
    print("An error occurred. See message above.")
